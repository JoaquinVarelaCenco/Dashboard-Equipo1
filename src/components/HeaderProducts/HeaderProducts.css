/* HEADER PRODUCTS */
.search-container__btnImage{
    opacity: .6;
}

.headerProducts{
    display: flex;
    width: 100%;
    justify-content: flex-end;
}

.header__search-container{
    display: flex;
    align-items: center;
    position: relative;
    width: 42px;
  }

  .header__search-container .hideComponent{
    display: none;
}

.header__search-container .showComponent{
    display: block;
    animation: fadeIn 3.5s;
}

@keyframes fadeIn {
    from{
        opacity: 0;
    }to{
        opacity: 1;
    }
}

  .search-container__btnClose{
    display: none;
    height: 32px;
    width: 25px;
    border-radius: 50%;
    margin-right: 5px;
    font-family: 'Roboto', sans-serif;
    font-size: 16px;
    font-weight: 700;
    background-color: transparent;
    color: white;
  }
  
.header__search-container input{
    border: 1px solid rgba(0, 0, 0, 0.5);
    border-radius: 16px;
    height: 32px;
    width: 100%;
    padding: 0 0 0 20px;
}
 
.header__search-container input:placeholder{
    display: none;
}

.search-container__btnSearch{
    position: absolute;
    right: 8px;
    background: none;
    border: none;
}

.expandSearchBarStyle .search-container__btnSearch{
    animation: fadeIn 3s;
}

.headerProducts-ContainerAgregar{
    align-self: center;
}

.headerProducts-ContainerAgregar a{
    text-decoration: none;
}
.headerProducts__btnAgregar{
    display: none;
    height: 32px;
    width: 150px;
    padding: 8px;
    align-self: center;
    border: 1px solid rgba(0, 0, 0, 0.5);
    border-radius: 16px;
    margin-left: 10px;
    text-decoration: none;
}

.headerProducts__btnAgregarAlternative{
    display: block;
    height: 32px;
    width: 30px;
    align-self: center;
    border: 1px solid rgba(0, 0, 0, 0.5);
    border-radius: 50%;
    margin-left: 10px;
    text-decoration: none;
    background-color: white;
    font-size: 25px;
    font-weight: 700;
}


@media (min-width: 1025px){
    .headerProducts__btnAgregar{
        display: block;
    }

    .headerProducts__btnAgregarAlternative{
        display: none;
    }

    .header__search-container{
        width: 280px;
    }
    .header__search-container input:placeholder{
        display: block;
    }
}

@media (min-width: 501px){
    .header__search-container{
        width: 280px;
    }

}


.headerProducts .expandSearchBarStyle{
    width: 100%;
    transition: width, 1s;
}

.headerProducts .expandSearchBarStyle:hover{
    width: 100%;
}

.headerProducts .closeSearchBarAnimation{
    transition: width, 1s;
}

.headerProducts .closeSearchBarAnimation:hover{
    width: 42px;
}



