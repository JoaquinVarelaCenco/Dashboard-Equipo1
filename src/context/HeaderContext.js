import { useState, createContext } from "react";

export const HeaderContext = createContext("home");

export const HeaderProvider = ({ children }) => {
  const [page, setPage] = useState("home");

  const value = {
    page,
    currentPage: (p) => setPage(p),
  };

  return (
    <HeaderContext.Provider value={value}>{children}</HeaderContext.Provider>
  );
};
