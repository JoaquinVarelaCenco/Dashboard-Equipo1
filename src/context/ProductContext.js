import { useState, createContext } from "react";

export const ProductContext = createContext("home");

export const ProductProvider = ({ children }) => {
  const defaultValues = {
    title: "",
    price: 0,
    stock: 0,
    description: "",
    category: "",
    images: [],
    rating: {
      rate: 0,
      count: 0,
    },
  };

  const [product, setProduct] = useState(defaultValues);

  const [lastState, setLastState] = useState(defaultValues);

  const value = {
    product,
    currentProduct: (p) => setProduct(p),
    lastState,
    currentLastProduct: (i) => setLastState(i),
    resetCamps: () => setProduct(lastState),
    resetForm: () => {
      setProduct(defaultValues);
      setLastState(defaultValues);
    },
  };

  return (
    <ProductContext.Provider value={value}>{children}</ProductContext.Provider>
  );
};
