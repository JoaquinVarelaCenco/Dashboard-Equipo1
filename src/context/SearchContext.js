import { useState, createContext, useEffect } from "react";
import { getProducts } from "../services/apiServices";
import { orderByAlphabet, orderByHighPrice, orderByLowPrice, orderByMostRelevants } from "../utils/product";

export const SearchContext = createContext();

export const SearchProvider = ({ children }) => {
  const [products, setProducts] = useState([]);
  const [productsExist, setProductsExist] = useState(true);
  const [helpProducts, setHelpProducts] = useState([]);
  const [searchTermValue, setSearchTermValue] = useState("");
  const [orderBy, setOrderBy] = useState("Mas relevantes");

  const getAllProducts = () => {
    getProducts()
      .then((data) => {
        setProductsExist(true);
        setProducts(data);
        setHelpProducts(data);
        orderProductsInit(data);
      })
      .catch(() => {
        setProductsExist(false);
      });
  };

  const filterProducts = (searchTerm) => {
    const newProducts = helpProducts.filter((val) => {
      if (
        searchTerm === "" ||
        val.title.toLowerCase().includes(searchTerm.toLowerCase())
      ) {
        return val;
      } else if (
        val.description.toLowerCase().includes(searchTerm.toLowerCase())
      ) {
        return val;
      }
    });
    return newProducts;
  };

  const handleSearch = (e) => {
    const newFilter = filterProducts(e.target.value);
    setSearchTermValue(e.target.value);
    setProducts(newFilter);
  }

  const orderProducts = (e) => {
    switch (e.target.value) {
      case "Mayor precio":
        setOrderBy("Mayor precio");
        setProducts(orderByHighPrice(products));
        break;
      case "Menor precio":
        setOrderBy("Menor precio");
        setProducts(orderByLowPrice(products));
        break;
      case "Mas relevantes":
        setOrderBy("Mas relevantes");
        setProducts(orderByMostRelevants(products));
        break;
      case "A-Z":
        setOrderBy("A-Z");
        setProducts(orderByAlphabet(products));
        break;
      default:
        setOrderBy("Mas relevantes");
        setProducts(orderByMostRelevants(products));
        break;
    }
  };

  const orderProductsInit = (data) => {
    switch (orderBy) {
      case "Mayor precio":
        setOrderBy("Mayor precio");
        setProducts(orderByHighPrice(data));
        break;
      case "Menor precio":
        setOrderBy("Menor precio");
        setProducts(orderByLowPrice(data));
        break;
      case "Mas relevantes":
        setOrderBy("Mas relevantes");
        setProducts(orderByMostRelevants(data));
        break;
      case "A-Z":
        setOrderBy("A-Z");
        setProducts(orderByAlphabet(data));
        break;
    }
  };

  useEffect(() => {
    getAllProducts();
  }, []);

  const value = {
    handleSearch,
        products,
        productsExist,
        getAllProducts,
        helpProducts,
        searchTermValue,
        orderBy,
        orderProducts,
  }
  return (
    <SearchContext.Provider
      value={value}
    >
      {children}
    </SearchContext.Provider>
  );
};
