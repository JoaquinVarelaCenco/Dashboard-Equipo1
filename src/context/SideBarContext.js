import { useState, createContext } from 'react'

export const SideBarContext = createContext('unhide');

export const SideBarProvider = ({children}) => {

    const [ visibility, setVisibility ] = useState('unhide');

    const value = {
        visibility,
        toggleVisibility: () => setVisibility(prev => prev === 'unhide' ? 'hide' : 'unhide')
    }


return (
        <SideBarContext.Provider value={value}>
            {children}
        </SideBarContext.Provider>
    )
}