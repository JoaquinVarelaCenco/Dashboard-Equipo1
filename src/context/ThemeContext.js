import { useState, createContext } from 'react'

export const ThemeContext = createContext('clear');

export const ThemeProvider = ({children}) => {
    
    const initialMode = (localStorage.getItem("darkMode")) ? localStorage.getItem("darkMode") : 'clear';

    const [ theme, setTheme ] = useState(initialMode);

    const value = {
        theme,
        toggleTheme: () => setTheme(prev => prev === 'dark' ? 'clear' : 'dark')
    }
    localStorage.setItem("darkMode", theme);


return (
        <ThemeContext.Provider value={value}>
            {children}
        </ThemeContext.Provider>
    )
}