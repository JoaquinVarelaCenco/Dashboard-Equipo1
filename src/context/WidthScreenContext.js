import { useState, createContext } from 'react'

export const WidthScreenContext = createContext(window.innerWidth);

export const WidthScreenProvider = ({children}) => {
    const [ widthScreen, setWidthScreen ] = useState(window.innerWidth);

    const value = {
        widthScreen,
        updateWidthScreen: (newState) => setWidthScreen(newState)
    }

    window.addEventListener('resize', ()=>{
      setWidthScreen(window.innerWidth)
  })

  return (
    <WidthScreenContext.Provider value={value}>
      {children}
    </WidthScreenContext.Provider>
  )
}

