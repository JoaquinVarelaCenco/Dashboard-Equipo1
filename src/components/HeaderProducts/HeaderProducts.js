import { useRef, useEffect, useState, useContext } from "react";
import { Link } from "react-router-dom";
import { SearchContext } from "../../context/SearchContext";
import { HeaderContext } from "../../context/HeaderContext";
import { WidthScreenContext } from "../../context/WidthScreenContext";
import Button from "../Button/Button";
import "./HeaderProducts.css";
import searchImage from "../../assets/images/magnify.svg";

const HeaderProducts = () => {
  //Context
  const context = useContext(SearchContext);
  const { currentTitleContainer } = useContext(HeaderContext);
  const { widthScreen } = useContext(WidthScreenContext);

  //Estados para cambiar estilos
  const [styleSearchAnimation, setStyleSearchAnimation] = useState("");
  const [styleDisplayNone, setStyleDisplayNone] = useState("");
  const [styleCloseSearchAnimation, setStyleCloseSearchAnimation] =
    useState("");

  //Elementos del DOM Logica expandir input de búsqueda
  const inputSearch = useRef("");
  const inputSearchContainer = useRef("");
  const btnClose = useRef("");
  const containerAddProduct = useRef("");

  useEffect(() => {
    currentTitleContainer("displayFlex");
  }, []);

  useEffect(() => {
    const handleResize = () => {
      closeSearchInput();
      if (widthScreen > 500) {
        // inputSearch.current.style.width = "280px";
        setStyleCloseSearchAnimation("");
      }
    };
    handleResize();
  }, [widthScreen]);

  //Borramos placeholder de buscador en caso de que el componente se cargue en pantalla pequeña
  useEffect(() => {
    if (window.screen.width < 501) {
      inputSearch.current.placeholder = "";
    }
  }, [inputSearch]);

  //cambiamos estilos al expandir searchBar
  const expandSearchInput = () => {
    if (widthScreen <= 500) {
      inputSearch.current.placeholder = "Buscar productos...";

      //este elemento depende del evento || NO del mediaQuery -
      currentTitleContainer("displayNone");
      containerAddProduct.current.style.display = "none";

      //Estados para agregar clases al cerar search input (modificacion con css)
      setStyleSearchAnimation("expandSearchBarStyle");
      setStyleDisplayNone("showComponent");
      setStyleCloseSearchAnimation("");
    } else {
      currentTitleContainer("displayFlex");
    }
  };

  //cambiamos estilos al cerrar searchBar
  const closeSearchInput = () => {
    inputSearch.current.placeholder = "";

    if (widthScreen > 500) {
      inputSearch.current.placeholder = "Buscar productos...";
    }

    //Estados para agregar clases al cerar search input (modificacion con css)
    setStyleSearchAnimation("");
    setStyleDisplayNone("");
    setStyleDisplayNone("hideComponent");
    setStyleCloseSearchAnimation("closeSearchBarAnimation");
    setTimeout(() => {
      currentTitleContainer("displayFlex");
    }, 200);
    containerAddProduct.current.style.display = "block";
  };

  return (
    <div className="headerProducts">
      <div
        className={`headerProducts__search-container  ${styleSearchAnimation} ${styleCloseSearchAnimation} `}
        ref={inputSearchContainer}
      >
        <Button
          click={closeSearchInput}
          classN={`headerProducts-search-container__btnClose headerBtn ${styleDisplayNone} noHoverEffect`}
          title={"X"}
          reference={btnClose}
        />

        <input
          type="text"
          className={`header__search  ${styleSearchAnimation}`}
          placeholder="Buscar productos..."
          ref={inputSearch}
          onChange={context.handleSearch}
        />

        <Button
          click={expandSearchInput}
          classN={
            "headerProducts-search-container__btnSearch headerBtn noHoverEffect"
          }
        >
          <img
            src={searchImage}
            alt="Lupa de busqueda"
            className="headerProducts-search-container__btnImage"
          />
        </Button>
      </div>
      <div className="headerProducts-ContainerAdd">
        <Link to={"/products/new"}>
          <Button
            title={"Agregar Producto"}
            classN={"headerProducts__btnAdd"}
          />
        </Link>
      </div>
      <div className="headerProducts-ContainerAdd">
        <Link to={"/products/new"} ref={containerAddProduct}>
          <Button
            classN={"headerProducts__btnAddAlternative noHoverEffect"}
            title={"+"}
          />
        </Link>
      </div>
    </div>
  );
};

export default HeaderProducts;
