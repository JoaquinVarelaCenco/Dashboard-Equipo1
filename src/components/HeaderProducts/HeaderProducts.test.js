import { render, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { SearchContext } from "../../context/SearchContext";
import HeaderProducts from './HeaderProducts';
import {BrowserRouter} from 'react-router-dom'
import { HeaderContext } from "../../context/HeaderContext";
import { WidthScreenContext } from "../../context/WidthScreenContext";


function renderHeaderProducts(handleSearch, widthScreen) {
    return render(
        <WidthScreenContext.Provider value={
            widthScreen
        } >
        <HeaderContext.Provider value={
            {currentTitleContainer: ()=>"displayFlex"}
            }
        >
            <SearchContext.Provider value={
                handleSearch
            }>
                <HeaderProducts/>
            </SearchContext.Provider>
        </HeaderContext.Provider>
        </WidthScreenContext.Provider>,
     { wrapper: BrowserRouter });
  }
    let btnExpand;

describe("Test header de productos", ()=>{
let component;

    beforeEach(()=>{
        let handleSearch = ()=>"";
        let width = {widthScreen: 400}
        component = renderHeaderProducts(handleSearch, width)
        btnExpand = document.querySelector('.headerProducts-search-container__btnSearch');
        
    })

    test("Se debe renderizar header products", ()=>{
        const { container } = component
        expect(container).toMatchSnapshot()
    })

    test("Se debe expandir inputSearch al presionar sobre el botón de la lupa", ()=>{

        //verificamos que el placeholder se muestre vacío en un principio
        expect(inputSearch.getAttribute('placeholder')).toBe("")
        
        userEvent.click(btnExpand);
        //verificamos que se muestre placeholder
        expect(inputSearch.getAttribute('placeholder')).toBe("Buscar productos...")

    })

    test("Se debe colapsar inputSearch al presionar sobre el botón X", ()=>{
        let btnColapse = screen.getByRole('button', {name: "X"})

        //expandimos input para que se muestre el placeHolder
        userEvent.click(btnExpand)
        //colapsamos input para que se oculte el placeHolder
        userEvent.click(btnColapse);

        expect(inputSearch.getAttribute('placeholder')).toBe("")
    })


    test("Se debe poder escribir en el input de busqueda y su valor debe ser el ingresado", ()=>{
        let testWord = "test";

        userEvent.type(inputSearch, testWord);
        expect(inputSearch).toHaveValue(testWord);
    })

    test("Los dos botones de agregar producto deben redirigir a /products/new", ()=>{
        const btnsAgregarProducto = screen.getAllByRole('link')
        btnsAgregarProducto.forEach(btn =>{
            expect(btn.getAttribute('href')).toBe('/products/new');
        })
    })

})