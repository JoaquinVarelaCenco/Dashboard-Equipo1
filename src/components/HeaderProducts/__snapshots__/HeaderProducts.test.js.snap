// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test header de productos Se debe renderizar header products 1`] = `
<div>
  <div
    class="headerProducts"
  >
    <div
      class="headerProducts__search-container   closeSearchBarAnimation "
    >
      <button
        class="headerProducts-search-container__btnClose headerBtn hideComponent noHoverEffect"
      >
        X
      </button>
      <input
        class="header__search  "
        placeholder=""
        type="text"
      />
      <button
        class="headerProducts-search-container__btnSearch headerBtn noHoverEffect"
      >
        <img
          alt="Lupa de busqueda"
          class="headerProducts-search-container__btnImage"
          src="magnify.svg"
        />
      </button>
    </div>
    <div
      class="headerProducts-ContainerAdd"
    >
      <a
        href="/products/new"
      >
        <button
          class="headerProducts__btnAdd"
        >
          Agregar Producto
        </button>
      </a>
    </div>
    <div
      class="headerProducts-ContainerAdd"
    >
      <a
        href="/products/new"
        style="display: block;"
      >
        <button
          class="headerProducts__btnAddAlternative noHoverEffect"
        >
          +
        </button>
      </a>
    </div>
  </div>
</div>
`;
